Add way to check whether iframe's owner element may follow dark appearance

basuke opened the pull request:
#### f99fba6f4b88b7538876ecc2d490efc81eeebdb7
<pre>
Add way to check whether iframe&apos;s owner element may follow dark appearance
<a href="https://bugs.webkit.org/show_bug.cgi?id=286713">https://bugs.webkit.org/show_bug.cgi?id=286713</a>
<a href="https://rdar.apple.com/143827621">rdar://143827621</a>

Reviewed by Aditya Keerthi.

Need to detect iframe element in main document may respond to dark appearance in CSS or document’s meta tag.

Add `Document::resolvedColorScheme(RenderStyle*) const` to get the element&apos;s color scheme to check which scheme
is ready to be used for the element.

* LayoutTests/http/tests/iframe-monitor/dark-mode-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/dark-mode.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/monitor-setup.js:
(async waitUntilUnload):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolvedColorScheme const):
(WebCore::Document::useDarkAppearance const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::generateResourceMonitorErrorHTML):
(WebCore::LocalFrame::showResourceMonitoringError):
* Source/WebCore/rendering/style/RenderStyleConstants.h:

Canonical link: <a href="https://commits.webkit.org/289613@main">https://commits.webkit.org/289613@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4fa8895784ff949fe93aff1a1e4bf2e99b03143d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87435 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6948 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41807 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92297 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38174 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89486 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7330 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15113 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67544 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25279 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90437 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79120 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47885 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33514 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37290 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34377 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94182 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14600 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10713 "Found 2 new test failures: fonts/ligature.html imported/w3c/web-platform-tests/css/selectors/invalidation/fullscreen-pseudo-class-in-has.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76361 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14854 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74975 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75576 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18606 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19961 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18377 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7538 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14618 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19913 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14361 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17805 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16144 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9290,6 +9290,14 @@ float Document::deviceScaleFactor() const
 }
 
+#if ENABLE(DARK_MODE_CSS)
+OptionSet<ColorScheme> Document::colorScheme(const RenderStyle* style) const
```
I think we should call this `resolvedColorScheme`.

pxlcoder commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9290,6 +9290,14 @@ float Document::deviceScaleFactor() const
+#if ENABLE(DARK_MODE_CSS)
+OptionSet<ColorScheme> Document::colorScheme(const RenderStyle* style) const
+{
+    bool isNormal = !style || style->colorScheme().isNormal();
+    return isNormal ? m_colorScheme : style->colorScheme().colorScheme();
```
Can you update `Document::useDarkAppearance` to call this method, rather than duplicate the logic.

basuke replied:
Is it okay to ref-ing the document here()? I'm afraid for the performance penalty.

pxlcoder replied:
Not sure I understand. There shouldn't be any ref-ing required here, nor do I see any.

basuke replied:
Ah, I should use `CheckedPtr` instead!

basuke replied:
No, that's wrong.

basuke replied:
Never mind. Safer C++ doesn't complaint about it.

pxlcoder commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1406,19 +1407,41 @@ void LocalFrame::setScrollingMode(ScrollbarMode scrollingMode)
 #if ENABLE(CONTENT_EXTENSIONS)
 
-static String generateResourceMonitorErrorHTML()
+static String generateResourceMonitorErrorHTML(bool mayUseLightApparance, bool mayUseDarkApparance)
```
I think we should have this take an `OptionSet<ColorScheme>`. Which isn't possible at this moment since that enum is behind `#if ENABLE(DARK_MODE_CSS)`. But I think you can just remove that `#if`.

basuke replied:
Yeah, `#if ENABLE(DARK_MODE_CSS)` this is the reason I was away from using `ColorScheme`.

pxlcoder replied:
You can go ahead and remove the `#if ENABLE(DARK_MODE_CSS)` around it.

pxlcoder replied:
(appearance is also spelled incorrectly in quite a few places (Apparance), but that should go away once you start using `ColorScheme`)

pxlcoder commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1428,10 +1451,19 @@ void LocalFrame::showResourceMonitoringError()
-    if (RefPtr document = this->document())
-        document->addConsoleMessage(MessageSource::ContentBlocker, MessageLevel::Error, "Frame was unloaded because its network usage exceeded the limit."_s);
+#if ENABLE(DARK_MODE_CSS)
+    auto style = iframeElement->existingComputedStyle();
```
```suggestion
    CheckedPtr style = iframeElement->existingComputedStyle();
```

pxlcoder commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1406,19 +1407,41 @@ void LocalFrame::setScrollingMode(ScrollbarMode scrollingMode)
+#endif
 #else
-    return WEB_UI_STRING("This frame is hidden for using too many system resources.", "Description HTML for frame unloaded by ResourceMonitor");
+    const auto lightAndDdarkScheme = ":root { color-scheme: light dark } "_s;
```
```suggestion
    const auto lightAndDarkScheme = ":root { color-scheme: light dark } "_s;
```

pxlcoder commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1406,19 +1407,41 @@ void LocalFrame::setScrollingMode(ScrollbarMode scrollingMode)
+    const auto darkStyle = "div { color: white } "_s;
+    const auto empty = ""_s;
+
+    bool useBoth = mayUseDarkApparance && mayUseLightApparance;
```
`useBoth` could probably be more descriptive.

pxlcoder commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1406,19 +1407,41 @@ void LocalFrame::setScrollingMode(ScrollbarMode scrollingMode)
+    const auto lightAndDarkScheme = ":root { color-scheme: light dark } "_s;
+    const auto onlyDarkScheme = ":root { color-scheme: only dark } "_s;
+    const auto darkStyle = "div { color: white } "_s;
+    const auto empty = ""_s;
```
These can be `constexpr`.

pxlcoder commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1406,19 +1407,41 @@ void LocalFrame::setScrollingMode(ScrollbarMode scrollingMode)
 {
     RefPtr iframeElement = dynamicDowncast<HTMLIFrameElement>(ownerElement());
-    if (!iframeElement)
+    RefPtr protectedDocument = this->document();
```
I think `this->` is unnecessary.

cdumez replied:
Actually no, we want the variable to be named `document`, not `protectedDocument`. We don't have `protected` variables anymore outside lambda capture.

basuke committed:
Add way to check whether iframe's owner element may follow dark appearance
https://bugs.webkit.org/show_bug.cgi?id=286713
rdar://143827621

Reviewed by Aditya Keerthi.

Need to detect iframe element in main document may respond to dark appearance in CSS or document’s meta tag.

Add `Document::resolvedColorScheme(RenderStyle*) const` to get the element's color scheme to check which scheme
is ready to be used for the element.

* LayoutTests/http/tests/iframe-monitor/dark-mode-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/dark-mode.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/monitor-setup.js:
(async waitUntilUnload):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::resolvedColorScheme const):
(WebCore::Document::useDarkAppearance const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::generateResourceMonitorErrorHTML):
(WebCore::LocalFrame::showResourceMonitoringError):
* Source/WebCore/rendering/style/RenderStyleConstants.h:

Canonical link: https://commits.webkit.org/289613@main
